src/qoi/src/rayfire_mesh.C

```diff
@@ -336,5 +362,79 @@ namespace GRINS
     // no convergence
     return false;
   }
+  
+  
+  void RayfireMesh::refine(const libMesh::Elem* main_elem, libMesh::Elem* rayfire_elem)
+  {
+    // these nodes cannot change
+    libMesh::Node* start_node = rayfire_elem->get_node(0);
+    libMesh::Node* end_node   = rayfire_elem->get_node(1);
+
+    // remove unrefined elem from _mesh
+    _mesh->delete_elem(rayfire_elem);
+
+    // remove unrefined elem from _elem_id_map
+    std::map<libMesh::dof_id_type,libMesh::Elem*>::iterator it;
+    it = _elem_id_map.find(main_elem->id());
+    _elem_id_map.erase(it);
+
+    // find which child elem we start with
+    unsigned int i=0;
+    while(!(main_elem->child(i))->contains_point(*start_node))
+      i++;
+
+    // we found the starting element, so perform a the rayfire
+    // until we reach the stored end_node
+    libMesh::Point* start_point = start_node;
+    libMesh::Point* end_point = new libMesh::Point();
+
+    const libMesh::Elem* next_elem;
+    const libMesh::Elem* prev_elem = main_elem->child(i);
+    
+    // if prev_elem is INACTIVE, then more than one refinement
+    // has taken place between reinit() calls and will
+    // break this
+    libmesh_assert( prev_elem->refinement_flag() == libMesh::Elem::RefinementState::JUST_REFINED );
+
+    // There are _mesh->n_nodes()-1 number of nodes already in _mesh
+    // so use n_nodes() as the ID of the next node to add
+    unsigned int end_node_id = _mesh->n_nodes();
+
+    unsigned int start_node_id = start_node->id();
+
+    // calculate the end point and
+    // get the second elem in the rayfire
+    next_elem = get_next_elem(prev_elem,start_point,end_point);
+
+    // iterate until we reach the stored end_node
+    while(!(end_point->absolute_fuzzy_equals(*end_node)))
+      {
+        // again, checking for multiple refinements
+        libmesh_assert( next_elem->refinement_flag() == libMesh::Elem::RefinementState::JUST_REFINED );
```

Comment:
Same here. `libmesh_assert()` should be used for things like checking the validity of a pointer. Otherwise, there are assert macros for conditional statements that print more informative messages.